### PR TITLE
8794 ref question

### PR DESCRIPTION
**Pull request by jpdimond**

### CODE REVIEW

File: app/models/question.rb [hunk 1]
```diff
@@ -276,4 +277,10 @@ def at_least_one_name
   def check_condition_integrity
     Condition.check_integrity_after_question_change(self)
   end
+
+  def valid_reference_url
+    return true if reference.blank?
```
Comment (smoyth) [inline app/models/question.rb]: Don't need the `true` here. You just need to not set an error.

File: app/models/question.rb [hunk 2]
```diff
@@ -276,4 +277,10 @@ def at_least_one_name
...
+    url = URI.parse(reference)
+    errors.add(:base, :valid_reference_url) unless url.is_a?(URI::HTTP) || url.is_a?(URI::HTTPS)
```
Comment (smoyth) [inline app/models/question.rb]: Nice

File: app/views/questions/_fields.html.erb [hunk 3]
```diff
@@ -9,32 +9,31 @@
...
-    <%= f.field(:audio_prompt, type: :file) %>
-  </div>
+  <%= f.field(:reference, type: :text) %>
+  <%= f.field(:audio_prompt, type: :file) %>
 
   <div class="minmax">
-    <%= f.field(:casted_minimum, :partial => 'questions/min_max_condition') %>
-    <%= f.field(:casted_maximum, :partial => 'questions/min_max_condition') %>
+    <%= f.field(:casted_minimum, partial: 'questions/min_max_condition') %>
+    <%= f.field(:casted_maximum, partial: 'questions/min_max_condition') %>
```
Comment (smoyth) [inline app/views/questions/_fields.html.erb]: Yay cleanups!

File: app/views/questions/_option_set_modal.html.erb [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<%# new option set modal %>
+<% if form_mode != :show && can?(:create, OptionSet) %>
+  <div class="new_option_set_form_wrapper" title="<%= I18n.t('activerecord.action_links.option_set.new') %>">
```
Comment (smoyth) [inline app/views/questions/_option_set_modal.html.erb]: Change single quotes to double in this file plz?

File: app/views/questions/_fields.html.erb (hunk 3 above)
Comment (smoyth) [reply app/views/questions/_fields.html.erb]: Can you also change single quotes to double?

File: db/migrate/20180822145730_add_reference_to_question.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddReferenceToQuestion < ActiveRecord::Migration[5.1]
```
[the remark this answers is not in the document]
Comment (smoyth) [reply db/migrate/20180822145730_add_reference_to_question.rb]: Please add the `frozen_string_literal: true` thingy.

File: spec/models/question_spec.rb [hunk 6]
```diff
@@ -93,6 +93,24 @@
...
+        let(:ref) { "https://eos.cartercenter.org/parts/6" }
+        it "has a valid reference" do
+          expect(question).to be_valid
+        end
+      end
+
+      context "with invalid reference" do
+        let(:ref) { "www.blah" }
+        it "has an invalid reference" do
+          expect(question).not_to be_valid
```
Comment (smoyth) [inline spec/models/question_spec.rb]: Should always check that you have the expected validation message. Else you could have a false negative. See below for the standard way to do that.

File: spec/models/question_spec.rb [hunk 7]
```diff
@@ -93,6 +93,24 @@
   end
 
   describe "validations" do
+    describe "reference url" do
+      let(:question) { build(:question, reference: ref) }
+
+      context "with valid reference" do
+        let(:ref) { "https://eos.cartercenter.org/parts/6" }
```
Comment (smoyth) [inline spec/models/question_spec.rb]: Should also test an http:// one?

File: config/locales/en/main.yml [hunk 8]
```diff
@@ -798,6 +799,7 @@ en:
           cant_delete_if_has_answers: "You can't delete this Question because it is referred to in at least one Answer."
           cant_delete_if_published: "You can't delete this Question because it appears on at least one published Form."
           general: "Question is invalid (see below)."
+          valid_reference_url: "Your reference link must begin with https or http"
```
Comment (smoyth) [inline config/locales/en/main.yml]: Please terminate with a period.

File: spec/models/question_spec.rb [hunk 9]
```diff
@@ -93,6 +93,24 @@
...
+        end
+      end
+
+      context "with invalid reference" do
+        let(:ref) { "www.blah" }
+        it "has an invalid reference" do
+          expect(question).not_to be_valid
+        end
+      end
+    end
```
Comment (smoyth) [inline spec/models/question_spec.rb]: Nice spec!

**Comment by smoyth** [comment]
Oh also, this field should probably have a hint on it...

File: config/locales/en/main.yml [hunk 10]
```diff
@@ -647,6 +647,7 @@ en:
         qtype_name: "The kind of information the user will be entering."
         access_level: "Inherit: Same access level as parent form. Private: No one can access the data via the API, regardless of parent form access level."
         metadata_type: "A special value that can be pre-filled into this question. If chosen, this question will be automatically hidden and not required, and any conditions will be removed."
+        reference: "A url that supports the question, must start with http or https."
```
Comment (smoyth) [inline config/locales/en/main.yml]: Slight edits: `A URL that supports the question. Must start with http:// or https://.`

File: app/views/questions/_fields.html.erb (hunk 3 above)
Comment (jpdimond) [reply app/views/questions/_fields.html.erb]: done

File: app/views/questions/_option_set_modal.html.erb (hunk 4 above)
Comment (jpdimond) [reply app/views/questions/_option_set_modal.html.erb]: done

**Pull request merged**